src/bookshelf/src/bksw_sec.erl

```diff
@@ -20,125 +21,159 @@
 
 -module(bksw_sec).
 
--export([is_authorized/2]).
+-export([encode_access_denied_error_response/3]).
+-export([is_authorized/2                      ]).
+-export([parse_authorization/1                ]).
 
--define(SECONDS_AT_EPOCH, 62167219200).
+-ifdef(TEST).
+-compile([export_all, nowarn_export_all       ]).
+-endif.
+
+% is this necessary?  try removing.
 -include("internal.hrl").
 
+-include_lib("erlcloud/include/erlcloud_aws.hrl").
+
 %%===================================================================
 %% API functions
 %%===================================================================
-is_authorized(Req0, #context{auth_check_disabled=true} = Context) ->
-    {true, Req0, Context};
-is_authorized(Req0, #context{} = Context) ->
-    Headers = mochiweb_headers:to_list(wrq:req_headers(Req0)),
-    {RequestId, Req1} = bksw_req:with_amz_request_id(Req0),
-    case proplists:get_value('Authorization', Headers, undefined) of
-        undefined ->
-            do_signed_url_authorization(RequestId, Req1, Context);
-        IncomingAuth ->
-            do_standard_authorization(RequestId, IncomingAuth, Req1, Context)
+
+is_authorized(Req0, #context{auth_check_disabled = true          } = Context) -> {true, Req0, Context};
+is_authorized(Req0, #context{auth_type           = presigned_url,
+                             date                = Date,
+                             incoming_sig        = IncomingSignature,
+                             signed_headers      = SignedHeaders,
+                             x_amz_expires_int   = XAmzExpiresInt} = Context) ->
+    Auth               = auth_init(Req0, Context, SignedHeaders),
+    {Bucketname, Key } = get_bucket_key(path(Auth)),
+    ComparisonURL      = mini_s3:s3_url(method(Auth), Bucketname, Key, XAmzExpiresInt, SignedHeaders, Date, config(Auth)),
+    IncomingSig        = list_to_binary(IncomingSignature),
+    [_, ComparisonSig] = string:split(ComparisonURL, "&X-Amz-Signature=", all),
+
+    % If the signature computation and comparison fails, this
+    % implementation attempts an alternative signature computation/comparison
+    % which adds or removes the port from the host header depending on whether
+    % the port is present or absent. At the time this was put in, there were
+    % problems with signature failures due to inconsistent treatment of host
+    % headers by various chef clients (present or missing ports).  Determining
+    % whether this capability is still necessary would require some investigation
+    % and testing.
+
+    CalculatedSig =
+        case IncomingSig of
+            ComparisonSig ->
```

Comment:
Note we should use a constant time comparison here for safety

Reply:
(Just making notes as to what we decided a few days ago)

Triaged as a 'nice to have' but not a merge blocker.